singlestoredb/apps/_cloud_functions.py

```diff
@@ -0,0 +1,63 @@
+import asyncio
+import typing
+import urllib.parse
+
+from ._config import AppConfig
+from ._process import kill_process_by_port
+
+if typing.TYPE_CHECKING:
+    from fastapi import FastAPI
+    from ._uvicorn_util import AwaitableUvicornServer
+
+# Keep track of currently running server
+_running_server: 'typing.Optional[AwaitableUvicornServer]' = None
+
+
+async def run_cloud_function(
+    app: 'FastAPI',
+    kill_existing_server: bool = True,
+    host: str = '0.0.0.0',
+    log_level: str = 'error',
+) -> None:
+    global _running_server
+    from ._uvicorn_util import AwaitableUvicornServer
+
+    try:
+        import uvicorn
+    except ImportError:
+        raise ImportError('package uvicorn is required to run cloud functions')
+
+    app_config = AppConfig.from_env()
+
+    if kill_existing_server:
```

Comment:
This wont work if i first started a dashboard server and then tried to start a functions server right?

Reply:
https://github.com/singlestore-labs/singlestoredb-python/pull/35/files/c47e7af5a961caaba8e930f06293c4d9abb6bc97#diff-048434818c0e36021286881716d67838897c42d6b66f41d611afded9cec0c54eR33-R41

So we first shutdown the functions server and after that kill any server that might be consuming the port by running `kill_process_by_port(...)`. So it'll cover both the cases. Only reason it is special case for functions is because functions server runs in the same process as kernel, so we have to shut it gracefully.